src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -97,14 +97,35 @@ HeapWord* ShenandoahFreeSet::allocate_single(ShenandoahAllocRequest& req, bool&
 
   // Overwrite with non-zero (non-NULL) values only if necessary for allocation bookkeeping.
 
+  bool allow_new_region = true;
+  switch (req.affiliation()) {
+    case ShenandoahRegionAffiliation::OLD_GENERATION:
+      if (_heap->old_generation()->adjusted_unaffiliated_regions() <= 0) {
+        allow_new_region = false;
+      }
+      break;
+
+    case ShenandoahRegionAffiliation::YOUNG_GENERATION:
+      if (_heap->young_generation()->adjusted_unaffiliated_regions() <= 0) {
+        allow_new_region = false;
+      }
+      break;
+
+    case ShenandoahRegionAffiliation::FREE:
+    default:
+      ShouldNotReachHere();
+      break;
+  }
+
   switch (req.type()) {
     case ShenandoahAllocRequest::_alloc_tlab:
     case ShenandoahAllocRequest::_alloc_shared: {
       // Try to allocate in the mutator view
       for (size_t idx = _mutator_leftmost; idx <= _mutator_rightmost; idx++) {
-        if (is_mutator_free(idx)) {
+        ShenandoahHeapRegion* r = _heap->get_region(idx);
+        if (is_mutator_free(idx) && (allow_new_region || r->affiliation() != ShenandoahRegionAffiliation::FREE)) {
```

Comment:
There is no concept of an allocation cursor per mutator.  In tracing some "anomalous" behaviors, I observed that the search for a heap region with memory available to be allocated can be very cumbersome.  As young memory becomes more scarce, the effort consumed by each thread trying to allocate (under lock by the way) becomes more and more costly, having to sequentially examine large numbers of regions (possibly more than a thousand regions) to find the first region with sufficient space to satisfy the allocation request.  We could definitely make some improvements here, especially because the allocating threads holds the lock throughout this traversal.  Another possible improvement is to not require the global heap lock while searching for a region to serve tlab allocation request.